### PR TITLE
Challenge rules

### PR DESCRIPTION
**Pull request by domino14**

### CODE REVIEW

File: game/game.go [hunk 1]
```diff
@@ -260,24 +271,156 @@ func (g *Game) validateTilePlayMove(m *move.Move) ([]alphabet.MachineWord, error
...
+		}
+
+	case pb.ChallengeRule_SINGLE:
+		if playLegal {
+			// There is no bonus score. Append a score of 0, however,
+			// to make it clear that we challenged.
+			lastTurn.Events = append(lastTurn.Events, bonusScoreEvent(0))
+		} else {
+			lastTurn.Events = []*pb.GameEvent{lastTurn.Events[0], offBoardEvent}
+		}
```
Comment (domino14) [inline game/game.go:381]: refactor this a bit ; the `!playLegal` event is the exact same in all cases.

File: game/challenge.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	lastTurn := g.history.Turns[len(g.history.Turns)-1]
+	cumeScoreBeforeChallenge := lastTurn.Events[0].Cumulative
+
+	offBoardEvent := &pb.GameEvent{
+		Nickname:   lastTurn.Events[0].Nickname,
+		Type:       pb.GameEvent_PHONY_TILES_RETURNED,
+		LostScore:  -lastTurn.Events[0].Score,
+		Cumulative: cumeScoreBeforeChallenge,
+		Rack:       lastTurn.Events[0].Rack,
+	}
```
Comment (jvc56) [inline game/challenge.go:48]: Can this be moved to line 63?

File: game/challenge.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+	bonusScoreEvent := func(bonus int32) *pb.GameEvent {
+		return &pb.GameEvent{
+			Nickname:   lastTurn.Events[0].Nickname,
+			Type:       pb.GameEvent_CHALLENGE_BONUS,
+			Bonus:      bonus + int32(addlBonus),
+			Cumulative: cumeScoreBeforeChallenge + bonus,
+		}
+	}
```
Comment (jvc56) [inline game/challenge.go:57]: Can this be moved to line 107?

File: game/game.go [hunk 4]
```diff
@@ -296,22 +360,42 @@ func (g *Game) PlayMove(m *move.Move, backup bool, addToHistory bool) error {
 
 		if addToHistory {
 			turn.Events = append(turn.Events, g.EventFromMove(m))
+			if g.history.LastKnownRacks == nil {
+				g.history.LastKnownRacks = []string{"", ""}
+			}
```
Comment (jvc56) [inline game/game.go:365]: This is janky. Why aren't we always setting the LastKnowRacks for every MoveTypePlay? I guess I don't know why they would even be nil.

File: game/game.go [hunk 5]
```diff
@@ -325,13 +409,36 @@ func (g *Game) PlayMove(m *move.Move, backup bool, addToHistory bool) error {
 		g.scorelessTurns++
 		if addToHistory {
 			turn.Events = append(turn.Events, g.EventFromMove(m))
+			if g.history.LastKnownRacks == nil {
+				g.history.LastKnownRacks = []string{"", ""}
+			}
```
Comment (jvc56) [inline game/game.go:414]: This is janky. Why aren't we always setting the LastKnowRacks for every MoveTypeExchange? I guess I don't know why they would even be nil.

**Pull request merged**